### PR TITLE
gdal: Add hdf4 support

### PR DESCRIPTION
**Pull request by knedlsepp**
###### Motivation for this change
I want to add hdf4 support to the gdal library.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/gdal/default.nix [hunk 1]
```diff
@@ -41,6 +46,7 @@ composableDerivation.composableDerivation {} (fixed: rec {
     "--with-python"               # optional
     "--with-static-proj4=${proj}" # optional
     "--with-geos=${geos}/bin/geos-config"# optional
+    "--with-hdf4=${hdf4}"
```
Comment (FRidh) [inline pkgs/development/libraries/gdal/default.nix]: can you add the note that this is optional

**Review by FRidh — CHANGES_REQUESTED** [review]
Perhaps its worth to separate hdf4 into multiple outputs.

File: pkgs/development/libraries/gdal/default.nix (hunk 1 above)
Comment (knedlsepp) [inline pkgs/development/libraries/gdal/default.nix]: Done. I'll also look into the multiple outputs issue.

**Comment by knedlsepp** [comment]
@FRidh I split hdf4 up into bin/dev/out. I had to manually specify postInstall `moveToOutput bin "$bin"` however, because otherwise the build would fail. Not exactly sure if this is the correct way to do it.

**Comment by knedlsepp** [comment]
@FRidh: Would you consider merging?

**Comment by FRidh** [comment]
```
The following tests FAILED:
        127 - HDF_TEST-testhdf-shared (Failed)
        128 - HDF_TEST-buffer-shared (Failed)
        133 - MFHDF_TEST-cdftest-shared (Failed)
        134 - MFHDF_TEST-hdfnctest-shared (Failed)
Errors while running CTest

builder for ‘/nix/store/cmjc162yw5haxd0qws7r2c12hlbvw3za-hdf-4.2.12.drv’ failed with exit code 8
error: build of ‘/nix/store/cmjc162yw5haxd0qws7r2c12hlbvw3za-hdf-4.2.12.drv’ failed
```
Some of the errors that I could find in the log:
```
/build/hdf-4.2.12/build/bin/testhdf-shared: error while loading shared libraries: libhdf.so.4.2.12: cannot open shared object file: No such file or directory
/build/hdf-4.2.12/build/bin/hdfnctest-shared: error while loading shared libraries: libmfhdf.so.4.2.12: cannot open shared object file: No such file or directory
```
I think you might want to undo the multiple outputs. I tried to use them with hdf5 as well after I suggested it to you, but it didn't work out at all.

**Comment by FRidh** [comment]
Okay, so I tried myself without multiple outputs, and I still get the errors.

**Comment by FRidh** [comment]
Do you have a NixOS machine with sandbox that you could test on?

**Comment by knedlsepp** [comment]
Oh. Didn't expect that the NixOS build would fail. I'll spin up a VM and see what's the deal. I suppose it will be the same problem as on the mac, where I need to set $DYLD_LIBRARY_PATH for the checkPhase. The test-binaries don't seem to know that they should look for the library in the build directory.

**Comment by knedlsepp** [comment]
Adding an "export LD_LIBRARY_PATH=$(pwd)/bin" seemed to help.

**Comment by knedlsepp** [comment]
@FRidh: Should work by now.

**Review by 0xABAB — APPROVED** [review]

**Pull request merged**